src/custom-element.js

```diff
@@ -863,13 +866,34 @@ function createBaseCustomElementClass(win) {
 
     /**
      * Called when the element is disconnected from the DOM.
+     *
+     * This callback is guarded by checks to see if the element is still
+     * connected. See #12849, https://crbug.com/821195, and
+     * https://bugs.webkit.org/show_bug.cgi?id=180940.
+     *
      * @final @this {!Element}
      */
     disconnectedCallback() {
-      if (this.isInTemplate_) {
+      // Short circut: Only call DOM methods if we think we're connected.
+      if (this.isConnected_ && dom.isConnectedNode(this)) {
         return;
       }
-      if (!this.isConnected_ || dom.isConnectedNode(this)) {
+      this.disconnect();
+    }
+
+    /** The Custom Elements V0 sibling to `disconnectedCallback`. */
+    detachedCallback() {
+      this.disconnectedCallback();
+    }
+
+    /**
+     * Called when an element is disconnected from DOM, or when an ampDoc is
+     * being disconnected (the element itself may still be connected to ampDoc).
+     *
+     * This does not guard against connected elements.
```

Comment:
This is confusing since we still check `this.isConnected_`. Maybe rename that variable to `connectedCallbackWasCalled_`?